### PR TITLE
[ADD] speedup sale migration by prepopulating computed fields

### PR DESCRIPTION
**Pull request by hbrunn**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale/migrations/9.0.1.0/post-migration.py [hunk 1]
```diff
@@ -55,6 +55,22 @@ def set_crm_team_message_types(env):
...
+    env.cr.execute(
+        """select l.id
+        from sale_order_line l
+        join sale_order_line_invoice_rel r on r.order_line_id=l.id
+        join account_invoice_line il on r.invoice_line_id=il.id
+        where l.product_uom<>il.uom_id
+        """)
+    env['sale.order.line'].invalidate_cache(
+        fnames=['invoice_status', 'qty_invoiced', 'qty_to_invoice'],
+        ids=[x for x, in env.cr.fetchall()]
```
Comment (pedrobaeza) [inline addons/sale/migrations/9.0.1.0/post-migration.py]: Shouldn't this be `ids[x[0] for x in ...]`?

Comment (hbrunn) [reply addons/sale/migrations/9.0.1.0/post-migration.py]: no, look closely for the comma

Comment (pedrobaeza) [reply addons/sale/migrations/9.0.1.0/post-migration.py]: Ah, a bit weird syntax, but OK, you're right

Comment (hbrunn) [reply addons/sale/migrations/9.0.1.0/post-migration.py]: the weird syntax is actually slightly faster - compare

`python -m timeit '[x for x, _ in [(y, y) for y in range(100000)]]'`

and

`python -m timeit '[x[0] for x in [(y, y) for y in range(100000)]]'`

Comment (pedrobaeza) [reply addons/sale/migrations/9.0.1.0/post-migration.py]: Noted for future list comprehension that I make.

Comment (hbrunn) [reply addons/sale/migrations/9.0.1.0/post-migration.py]: this can also be helpful in a normal loop when you iterate over a huge result set, but do quite little per iteration. Then

`for a, b, c, d in cr.fetchall()`

performs better than

`for row in cr.fetchall()`

if within the loop, all values of `row` are used

Comment (pedrobaeza) [reply addons/sale/migrations/9.0.1.0/post-migration.py]: Thanks for the info. The problem is if the list if very huge or you don't know all possible values (in a `select *` statement), but obviously that cases can be always contained.

**Comment by NL66278** [comment]
It might speed things up even more if you first add all the fields to a table and then fill the values with one statement. Now you have to go several times to all or most of the sale orders and lines.

**Comment by hbrunn** [comment]
yes, good point.

But at the moment I'm busy with a general speedup of recomputations of which I promise myself much more performance gains, because something is fishy here. It runs currently for testing purposes, afterwards I'll propose this to upstream, OCB and here https://github.com/hbrunn/OCB/commit/25e5f9526f182d18dd37a4fbcea07f638bea2a92

There are two problems with current recomputation: The time used grows nonlinear, that's a problem with my 300k sale order lines. Additionally, what makes this hard to debug is that there's some nondeterminism going on, two times recomputing the same field on the same rows yields a different execution path (or the same, that's up to chance).

**Comment by NL66278** [comment]
Apart from the general rewrite of recomputation you might also look at this PR: https://github.com/OCA/OpenUpgrade/pull/735
Allows to just stop the expensive recomputations, while having Odoo create the fields themselves. Consequence is that the actual data_migration must move to the post_migration

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 2]
```diff
@@ -56,9 +56,83 @@ def map_order_state(cr):
...
+        )
+        update sale_order_line
+        set qty_invoiced=line2qty_invoiced.qty_invoiced
+        from line2qty_invoiced where line2qty_invoiced.id=sale_order_line.id
+        """)
+    cr.execute('alter table sale_order_line add column qty_to_invoice numeric')
+    cr.execute(
+        """\
+        update sale_order_line
+        set qty_to_invoice=product_uom_qty - qty_invoiced
```
Comment (JordiBForgeFlow) [inline addons/sale/migrations/9.0.1.0/pre-migration.py]: @hbrunn This sql sentence is not correct. It does not replicate the same conditions as https://github.com/odoo/odoo/blob/9.0/addons/sale/sale.py#L516

For example, in products where the invoice policy is set based on delivered quantities the quantity to invoice is to be based on the quantity delivered - quantity invoiced.

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 3]
```diff
@@ -56,9 +56,83 @@ def map_order_state(cr):
...
+        from sale_order o join product_pricelist p on o.pricelist_id=p.id
+        where l.order_id=o.id
+        """)
+    cr.execute('alter table sale_order_line add column invoice_status varchar')
+    cr.execute("update sale_order_line set invoice_status='no'")
+    cr.execute(
+        """\
+        update sale_order_line
+        set invoice_status=(
+            case when qty_to_invoice<=0 then 'invoiced'
```
Comment (JordiBForgeFlow) [inline addons/sale/migrations/9.0.1.0/pre-migration.py]: replace with `case when coalesce(qty_to_invoice,0)<=0 then 'invoiced'`

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 4]
```diff
@@ -56,9 +56,83 @@ def map_order_state(cr):
...
+        update sale_order_line
+        set qty_invoiced=line2qty_invoiced.qty_invoiced
+        from line2qty_invoiced where line2qty_invoiced.id=sale_order_line.id
+        """)
+    cr.execute('alter table sale_order_line add column qty_to_invoice numeric')
+    cr.execute(
+        """\
+        update sale_order_line
+        set qty_to_invoice=product_uom_qty - qty_invoiced
+        where qty_invoiced is not null
```
Comment (JordiBForgeFlow) [inline addons/sale/migrations/9.0.1.0/pre-migration.py]: This where condition is not correct. Instead use:
```
update sale_order_line
set qty_to_invoice=product_uom_qty - coalesce(qty_invoiced,0)
```

This issue has been giving me a hard time.

**Comment by hbrunn** [comment]
@jbeficent thanks for you input, done. Next time, just create a PR

**Comment by pedrobaeza** [comment]
What is the status of this?

**Comment by hbrunn** [comment]
I'm waiting for @jbeficent to test the last commit where I integrated his points

**Comment by pedrobaeza** [comment]
@jbeficent can you review? Travis is red anyway.

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 5]
```diff
@@ -56,9 +56,90 @@ def map_order_state(cr):
...
+        update sale_order_line
+        set qty_invoiced=line2qty_invoiced.qty_invoiced
+        from line2qty_invoiced where line2qty_invoiced.id=sale_order_line.id
+        """)
+    cr.execute('alter table sale_order_line add column qty_to_invoice numeric')
+    cr.execute(
+        """\
+        update sale_order_line
+        set qty_to_invoice=case
+            when pt.invoice_policy = 'order' then
```
Comment (omar7r) [inline addons/sale/migrations/9.0.1.0/pre-migration.py]: invoice_policy field in product.template doesn't exists in pre-migration.py because it's a new field in 9.0 version If you move this query to post-migrations it works

**Comment by pedrobaeza** [comment]
@jbeficent please feedback about this PR.

@hbrunn check last comment and red status

**Comment by hbrunn** [comment]
@pedrobaeza last comment defeats the purpose of the pr, the code is fixed but travis is acting up

**Comment by pedrobaeza** [comment]
Again the TRAVIS_COMMIT error, so I close and reopen the PR for forcing a correct build.

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 6]
```diff
@@ -56,9 +56,85 @@ def map_order_state(cr):
         WHERE sol.order_id = so.id""")
 
 
+def prepopulate_fields(cr):
+    """ Recomputing a fields will be very expensive via the ORM, so do it
+    here for fields where the computation is trivial"""
+    cr.execute('alter table sale_order_line add column qty_invoiced numeric')
```
Comment (pedrobaeza) [inline addons/sale/migrations/9.0.1.0/pre-migration.py:62]: ADD `set default=0`

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 7]
```diff
@@ -56,9 +56,85 @@ def map_order_state(cr):
...
+            join account_invoice i on i.id=il.invoice_id
+            join product_uom uom_invoice on il.uom_id=uom_invoice.id
+            join product_uom uom_sale on l.product_uom=uom_sale.id
+            group by l.id
+        )
+        update sale_order_line
+        set qty_invoiced=line2qty_invoiced.qty_invoiced
+        from line2qty_invoiced where line2qty_invoiced.id=sale_order_line.id
+        """)
+    cr.execute('alter table sale_order_line add column qty_to_invoice numeric')
```
Comment (pedrobaeza) [inline addons/sale/migrations/9.0.1.0/pre-migration.py:85]: ADD `set default=product_uom_qty`

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 8]
```diff
@@ -56,9 +56,85 @@ def map_order_state(cr):
...
+        """)
+    cr.execute('alter table sale_order_line add column currency_id integer')
+    cr.execute(
+        """\
+        update sale_order_line l set currency_id=p.currency_id
+        from sale_order o join product_pricelist p on o.pricelist_id=p.id
+        where l.order_id=o.id
+        """)
+    cr.execute('alter table sale_order_line add column invoice_status varchar')
+    cr.execute("update sale_order_line set invoice_status='no'")
```
Comment (pedrobaeza) [inline addons/sale/migrations/9.0.1.0/pre-migration.py:102]: Instead of this, just add in the previous statement `set default='no'`

File: addons/sale/migrations/9.0.1.0/pre-migration.py [hunk 9]
```diff
@@ -56,9 +56,85 @@ def map_order_state(cr):
...
+        """\
+        update sale_order_line
+        set invoice_status=(
+            case when coalesce(qty_to_invoice, 0)<=0 then 'invoiced'
+            else 'to invoice' end
+        )
+        where state in ('sale', 'done')
+        """)
+    cr.execute('alter table sale_order add column invoice_status varchar')
+    cr.execute("update sale_order set invoice_status='no'")
```
Comment (pedrobaeza) [inline addons/sale/migrations/9.0.1.0/pre-migration.py:113]: The same (`set default='no'`)

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by mvaled** [comment]
Should we merge this?

**Comment by mvaled** [comment]
I'm testing this today.  I'll be reporting later (tomorrow perhaps)

**Comment by mvaled** [comment]
TL;DR It seems this is incorrect.

### Experiment

I took a DB with Odoo 8, and clone it twice:

```
createdb -T o8 o9base
createdb -T o8 o9speedy
```

Then I migrated the `o9base` without this patch (https://github.com/merchise-autrement/openupgrade/tree/merchise-develop-9.0) and the `o9speedy` with the patch (https://github.com/merchise-autrement/OpenUpgrade/tree/merchise-9.0-sale-speedup).

However comparing the resulting `sale_order_line` and `sale_order` gives a lot of differences:

For instance, the field `sale.order.line.qty_invoiced` is different in almost all lines:

```
$ diff -u <(psql -d o9speedy -Ac 'select id, qty_invoiced from sale_order_line where order by id;') <(psql -d o9base -Ac 'select id, qty_invoiced from sale_order_line order by id;') | wc -l
191365
```

The first thing I notice is that where the `o9base` get 0, `o9speedy` gets NULL.  Maybe this is not a problem.  Because we can always do `UPDATE sale_order_line SET qty_invoiced=0 WHERE qty_invoiced IS NULL`.  Let's do it:

```
$ psql -d o9speedy -c "UPDATE sale_order_line SET qty_invoiced=0 WHERE qty_invoiced IS NULL"
UPDATE 62646
```

However, we still have a lot of differences:

```
$ diff -u <(psql -d o9speedy -Ac 'select id, qty_invoiced from sale_order_line order by id;') <(psql -d o9base -Ac 'select id, qty_invoiced from sale_order_line order by id;') | wc -l
82575
```

Let's see some records that differ:

```
-38|15.0000000000000000
-39|1.00000000000000000000
+38|0
+39|0
```

For the record 38, the speedier version computes 15 while the normal version computes 0.  Other fields show as much differences as well.  I'll try to look at these differences and see if I can spot the issue.

**Comment by mvaled** [comment]
I pushed 8625c1029ad5e5851a3f1e82f0a7fe01b8a14546 to fix the qty_invoiced amount.

**Comment by hbrunn** [comment]
can you PR this to this PR's branch? Then I'll merge this in this PR. If you want to speed up getting this merged, you can check @pedrobaeza's comments to. I won't have time to address them any time soon.

**Comment by mvaled** [comment]
I already did: https://github.com/hbrunn/OpenUpgrade/pull/4

**Comment by mvaled** [comment]
I'll be addressing @pedrobaeza's comments soon.

**Comment by mvaled** [comment]
See PR https://github.com/hbrunn/OpenUpgrade/pull/5

File: addons/sale/migrations/9.0.1.0/pre-migration.py (hunk 7 above)
Comment (mvaled) [reply addons/sale/migrations/9.0.1.0/pre-migration.py:85]: About this PG says: cannot use column references in default expression

Comment (mvaled) [reply addons/sale/migrations/9.0.1.0/pre-migration.py:85]: From PG doc:

> DEFAULT default_expr

> The DEFAULT clause assigns a default data value for the column whose column definition it appears within. The value is any *variable-free expression* (subqueries and cross-references to other columns in the current table are not allowed). The data type of the default expression must match the data type of the column.

> The default expression will be used in any insert operation that does not specify a value for the column. If there is no default for a column, then the default is null.

Comment (pedrobaeza) [reply addons/sale/migrations/9.0.1.0/pre-migration.py:85]: OK, I didn't know. Then we should still keep here 2 sentences

**Comment by mvaled** [comment]
I created another PR to the original branch.  It addresses @pedrobaeza's
requests.  One of the requested defaults cannot be set by SQL.

**Comment by mvaled** [comment]
Still working on this.  The field `qty_to_invoice` shows differences.

**Comment by mvaled** [comment]
Hum. The `qty_to_invoice` depends on the computation `qty_delivered`.  However `qty_delivered` is not computed in this addon but in `sale_analytic`.

File: addons/sale/migrations/9.0.1.0/pre-migration.py (hunk 7 above)
Comment (mvaled) [reply addons/sale/migrations/9.0.1.0/pre-migration.py:85]: I'm working on https://github.com/hbrunn/OpenUpgrade/pull/5.  I hope next week this can be merged.

**Comment by pedrobaeza** [comment]
Any news on this?

**Comment by mvaled** [comment]
I've been unable to keep working on this for the past weeks.  We're heavily busy migrating ~50 addons to Odoo 9 and then to Odoo 10.  After they all migrated to 9, I could try another round of DB migration and see if I could work on this.

**Comment by pedrobaeza** [comment]
OK, let me know. I don't see this as critical because times are more or less manageable in my cases, but maybe with a lot of sale orders it's not the same.

**Comment by pedrobaeza** [comment]
Closing this due to errors on the migration. If anyone wants to continue this experiment, can take this code and go on with the needed changes, proposing the new alternative.
